packages/api/src/router/resources.ts

```diff
@@ -429,7 +439,89 @@ export const resourceRouter = createTRPCRouter({
           .on({ type: "resource", id: input }),
     })
     .input(z.string().uuid())
-    .query(async ({ ctx, input }) => getNodesRecursively(ctx.db, input)),
+    .query(async ({ ctx, input }) => {
+      const resource = await ctx.db.query.resource.findFirst({
+        where: eq(schema.resource.id, input),
+      });
+      if (resource == null) return null;
+      const childrenNodes = await getNodesRecursively(ctx.db, input);
+
+      const fromNodesPromises = ctx.db
+        .select()
+        .from(schema.resourceRelationship)
+        .innerJoin(
+          schema.resource,
+          eq(
+            schema.resourceRelationship.fromIdentifier,
+            schema.resource.identifier,
+          ),
+        )
+        .where(
+          and(
+            eq(schema.resourceRelationship.workspaceId, resource.workspaceId),
+            eq(schema.resourceRelationship.toIdentifier, resource.identifier),
+          ),
+        )
+        .then((rows) =>
+          rows.map(async (row) => ({
+            ...row,
+            node: await getNodeDataForResource(ctx.db, row.resource.id),
+          })),
+        )
+        .then((promises) => Promise.all(promises));
+
+      const toNodesPromises = ctx.db
+        .select()
+        .from(schema.resourceRelationship)
+        .innerJoin(
+          schema.resource,
+          eq(
+            schema.resourceRelationship.toIdentifier,
+            schema.resource.identifier,
+          ),
+        )
+        .where(
+          and(
+            eq(schema.resourceRelationship.workspaceId, resource.workspaceId),
+            eq(schema.resourceRelationship.fromIdentifier, resource.identifier),
+          ),
+        )
+        .then((rows) =>
+          rows.map(async (row) => ({
+            ...row,
+            node: await getNodeDataForResource(ctx.db, row.resource.id),
+          })),
+        )
+        .then((promises) => Promise.all(promises));
+
+      const [fromNodes, toNodes] = await Promise.all([
+        fromNodesPromises,
+        toNodesPromises,
+      ]);
+
+      return {
+        resource,
+        nodes: [
+          ...childrenNodes,
+          ...fromNodes.map((n) => n.node),
+          ...toNodes.map((n) => n.node),
+        ].filter(isPresent),
+        associations: {
+          from: fromNodes
+            .filter((n) => n.node != null)
+            .map((n) => ({
+              ...n.resource_relationship,
+              resource: n.node!,
+            })),
```

Comment:
if you do isPresent check, typescript will know in the next map node is not null